Catch a syntax error when parsing an input file with the --stay-connected flag active

shaggysa opened the pull request:
Most errors are caught by the robot after the program has been sent, but a few will abort the entire program.
While this used to be fine, it can be annoying when the --stay-connected flag is active.

Currently, this only works after the program has been successfully sent once, but I'm open to applying it more broadly. However, getting that to work would probably require a small refactor.

<img width="703" height="253" alt="image" src="https://github.com/user-attachments/assets/a43f0454-581f-4fa4-a9bc-36dc0dd8b53e" />

shaggysa committed:
cli/__init__.py: catch a syntax error when re-compiling the target file

shaggysa committed:
cli/__init__.py: slightly modify the error message warning

shaggysa committed:
cli/__init__.py: refactor to allow the catching of syntax errors on the first download to the hub

shaggysa committed:
cli/__init__.py: clean up an if statement

shaggysa committed:
CHANGELOG.md: add an entry for the new changes to the run command

dlech commented on pybricksdev/cli/__init__.py:
```diff
@@ -375,6 +317,77 @@ async def reconnect_hub():
+            if args.stay_connected:
+                await stay_connected_menu(hub)
+
+        except SyntaxError as e:
```
I'm not sure we want this to have such a broad scope.

On the first run, if we get a syntax error, I think it is OK to exit immediately because this will occur before the connection is actually made.

Instead, I think we should only catch the SyntaxError when attempting to recompile after a connection was made.

shaggysa replied:
It does connect to the robot before attempting to send for the first time. Having the menu show up if the arg is present allows the user to easily adjust the file or change the target without having to re-run the command. I also feel like it's nice to keep the message from being buried in a stack trace.

dlech replied:
Ah, OK, I thought it was the other way around.

shaggysa committed:
cli/__init__.py: avoid using newline escape characters

shaggysa committed:
cli/__init__.py: move stay_connected_menu to separate function

shaggysa commented:
I moved it to a separate function, but I am unable to test it with a robot at the moment. The linter is happy, but I can't guarantee that it will work the way I intended.

dlech committed:
Merge branch 'master' into feature/catch-syntax-error

dlech commented:
Apparently I've done something to `mpy-cross` that makes it segfault instead of reporting a `SyntaxError`.

I made some other changes that I just merged to fix a different issue, but it means we no longer get the `SyntaxError` raised from `ModuleFinder`, so we will need to reconsider this a bit. But it looks like I will need to figure out what is going on with mpy-cross first and fix that.

dlech commented:
OK, the crash is fixed now (you will need to merge the `master` branch and then `poetry install` to get the updates).

Because of the other change I made, it will now raise `subprocess.CalledProcessError` instead of `SyntaxError`. We can catch this new exception and print the `stderr` instead of catching the `SyntaxError`.

It would also be nice if we could get some unit tests for this.

shaggysa committed:
explicitly pass args to the stay-connected menu from the run function

shaggysa committed:
Merge remote-tracking branch 'origin/feature/catch-syntax-error' into feature/catch-syntax-error

shaggysa committed:
Merge branch 'pybricks:master' into feature/catch-syntax-error

shaggysa committed:
catch a CalledProcessError and decode stderr

shaggysa committed:
fix typo in except block

shaggysa commented:
Ok, I think I'm catching the error properly now (I was actually able to test with a robot this time). The error seems to always show stdin as the file, which should probably change to make it a bit more helpful on multi-file projects.

shaggysa committed:
add a few unit tests related to stay_connected_menu

shaggysa committed:
make the open() call in the "Change Target File" option explicitly specify utf-8 encoding

shaggysa committed:
fix leaking coroutines in test_stay_connected_menu_interruptions

It looks like the mock menu had to have the side effect of an actual async function for it to return a proper coroutine. When its side effect was hardcoded values, the AsyncMock function was unable to be cancelled properly, causing coroutines to leak.

shaggysa commented:
I decided to go ahead and add program-cancelling from the terminal. I decided on a non-signal keypress because I didn't want to intercept a KBI in the case that the "racing" function was not active and I didn't want to worry about double presses cancelling everything. I briefly tried EOF, but it looked like I had to manually re-open stdin every time that EOF was received. 

I landed on an invisible PromptToolKit app that listens for the letter 'q'. The letter it listens for can be easily modified if desired.

This should at least partially address #67.

One quirk that I can't seem to figure out is that
```
{'pybricks.tools'}
{'pybricks.tools'}
```
is printed on every call to hub.download. I didn't touch that function, and I know that the race_keypress function isn't causing it because it doesn't happen on hub.start_user_program (which uses race_keypress) and it does happen on hub.download standalone (which doesn't use race_keypress).

dlech commented:
> I decided to go ahead and add program-cancelling from the terminal.

Sounds great. Can we save that for a separate pull request? One feature is enough for one pull request. :smile:

shaggysa committed:
add a test for the `Run Stored Program` option

The function that this option calls is not available on hub firmware versions prior to 3.2.0-beta.3, so this test confirms that it is not called when the hub's firmware version is too old.

shaggysa committed:
rearrange test order

shaggysa committed:
tweak mpy-cross error message and control flow

shaggysa committed:
create a syntax error organically in the unit test

shaggysa committed:
remove unused subprocess import

shaggysa committed:
un-mock the stay-connected menu in the syntax error test

shaggysa committed:
Merge branch 'pybricks:master' into feature/catch-syntax-error
